api/interop.go

```diff
@@ -18,53 +20,123 @@ import (
 func apiInteropHandler(w http.ResponseWriter, r *http.Request) {
 	ctx := shared.NewAppEngineContext(r)
 	passRateType := metrics.GetDatastoreKindName(metrics.PassRateMetadata{})
-	query := datastore.NewQuery(passRateType).Order("-StartTime").Limit(1)
+	query := datastore.NewQuery(passRateType).Order("-StartTime")
 
 	filters, err := shared.ParseTestRunFilterParams(r)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
 
-	// We 'load by SHA' by fetching any interop result with all TestRunIDs for that SHA.
+	// We load non-default queries by fetching any interop result with all their
+	// TestRunIDs present in the TestRuns matching the query.
+	var keysChecker func(shared.TestRunIDs) bool
 	if !filters.IsDefaultQuery() {
-		// Load default browser runs for SHA.
-		// Force any max-count to one; more than one of each product makes no sense for a interop run.
-		one := 1
-		filters.MaxCount = &one
-		runs, err := LoadTestRunsForFilters(ctx, filters)
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return
-		} else if len(runs) < 1 {
+		// When ?complete=true, make sure to show results for the same complete run
+		// (executed for all browsers). Because we don't want to mismatch an interop
+		// which has SHAs from 2 separated-but-complete runs, we need to keep the
+		// keys grouped.
+		if shared.IsLatest(filters.SHA) && filters.Aligned != nil && *filters.Aligned {
+			ten := 10
+			_, shaKeys, err := shared.GetAlignedRunSHAs(ctx, filters.GetProductsOrDefault(), filters.Labels, filters.From, filters.To, &ten)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			} else if len(shaKeys) < 1 {
+				// Bail out early - can't find any complete runs.
+				http.Error(w, "No metrics runs found", http.StatusNotFound)
+				return
+			}
+			keysChecker = checkKeysAreAligned(shaKeys)
+		} else {
+			// Load default browser runs for SHA.
+			// Force any max-count to one; more than one of each product makes no sense for a interop run.
+			shaFilters := filters
+			limit := 64
+			shaFilters.MaxCount = &limit
+			keys, err := LoadTestRunKeysForFilters(ctx, shaFilters)
+
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			} else if len(keys) < 1 {
+				http.Error(w, "No metrics runs found", http.StatusNotFound)
+				return
+			}
+			keysChecker = checkKeysMatchQuery(keys)
+		}
+	}
+
+	// Iterate until we find interop data where its TestRunIDs match the query.
+	var interop metrics.PassRateMetadata
+	it := query.Run(ctx)
```

Comment:
I'm not seeing any explicit limit on the query. In the worst case, how many iterations will we do?

Reply:
I'll cap to 1000. Though, interestingly, this points out a problem with loading really, really old interop data (e.g. for a very specific SHA).

Reply:
Let's make this 100.